src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+  for (auto& predicate_info : pull_up_info.predicates) {
+    // Note that we cannot remove the predicate nodes here just yet, because the reformulation might still fail when no
+    // valid primary join predicate is found.
+    join_predicates.emplace_back(std::make_shared<BinaryPredicateExpression>(
+        predicate_info.condition, predicate_info.left_operand, predicate_info.right_operand));
+  }
+
+  // Semi and anti joins are currently only implemented by hash joins. These need an equals comparison as the primary
+  // join predicate.
+  for (auto it = join_predicates.begin(), end = join_predicates.end(); it != end; ++it) {
+    const auto& predicate = **it;
+    if (predicate.predicate_condition == PredicateCondition::Equals) {
+      std::iter_swap(join_predicates.begin(), it);
+      break;
+    }
+  }
+
+  if (join_predicates.empty() || join_predicates.front()->predicate_condition != PredicateCondition::Equals) {
+    return _apply_to_inputs(node);
+  }
+
+  // Begin altering the LQP. All failure checks need to be finished at this point.
+  //
+  // Since multi-predicate joins are currently still work-in-progress, we emulate them by:
+  //   - Performing an inner join on the two trees using an equals predicate
+  //   - Putting the remaining predicates in predicate nodes above the join
+  //   - Inserting a projection above the predicates to filter out any columns from the right sub-tree
+  //   - Inserting a group by over all columns from the left subtree, to filter out duplicates introduced by the join
+  //
+  // NOTE: This only works correctly if the left sub-tree does not contain any duplicates. It also only works for
+  // emulating semi joins, not for anti joins. It is only meant to get the implementation started and to collect some
+  // preliminary benchmark results.
+  const auto left_columns = predicate_node->left_input()->column_expressions();
+  auto distinct_node = AggregateNode::make(left_columns, std::vector<std::shared_ptr<AbstractExpression>>{});
+  auto left_only_projection_node = ProjectionNode::make(left_columns);
+  const auto join_node = JoinNode::make(JoinMode::Inner, std::move(join_predicates.front()));
+  join_predicates.erase(join_predicates.begin());
+
+  lqp_replace_node(node, distinct_node);
+  lqp_insert_node(distinct_node, LQPInputSide::Left, left_only_projection_node);
+
+  std::shared_ptr<AbstractLQPNode> parent = left_only_projection_node;
+  for (auto& secondary_join_predicate : join_predicates) {
+    const auto secondary_predicate_node = PredicateNode::make(std::move(secondary_join_predicate));
+    lqp_insert_node(parent, LQPInputSide::Left, secondary_predicate_node);
+    parent = secondary_predicate_node;
+  }
+
+  lqp_insert_node(parent, LQPInputSide::Left, join_node);
+  join_node->set_right_input(right_tree_root);
+
+  // By removing/replacing nodes only now, the right input of the join node will be automatically kept up to date when
+  // we remove/replace the root of the previous subquery.
```

Comment:
If we remove the nodes before building the join node (which a previous version did), we need to keep track of the root of the right sub-tree. Doing it this way means that `lqp_remove_node` does that for us, since it is registered as an input to the join node. I admit the comment is hard to understand without the context, and maybe also unnecessary.

Reply:
If it just explains history that we don't know about, I'd remove it